src/edu/psu/sweng500/team8/gui/SudokuGUI.java

```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
+		
+		
+		final JFileChooser fc = new JFileChooser();
+
+		fc.setFileFilter(new FileFilter()
+        {
+           @Override
+           public boolean accept(File file)
+           {
+              return (file.getName().toUpperCase().endsWith(".SUDOKU") || file.isDirectory());
+           }
+
+           @Override
+           public String getDescription()
+           {
+              return "Sudoku files";
+           }
+        });
+		
+		fc.setAcceptAllFileFilterUsed(false);
+		
+		int returnVal = fc.showSaveDialog(getComponent(0));;
+		
+		if(returnVal == JFileChooser.APPROVE_OPTION){
+			
+
+				File f = fc.getSelectedFile();
+				String path = f.getAbsolutePath();
+				
+				if(f.exists()){
+		            int result = JOptionPane.showConfirmDialog(this,"The file exists, overwrite?","Existing file",JOptionPane.YES_NO_CANCEL_OPTION);
+		            switch(result){
+		                case JOptionPane.YES_OPTION:
+		                    fc.approveSelection();
+		                   // savePuzzle(path);
+		                    return;
+		                case JOptionPane.NO_OPTION:
+		                    return;
+		                case JOptionPane.CLOSED_OPTION:
+		                    return;
+		                case JOptionPane.CANCEL_OPTION:
+		                    fc.cancelSelection();
+		                    return;
+		            }
+		        }
+		        fc.approveSelection();
+		        savePuzzle(path);
+		        
+		        gameBoard.remarkGivenCells(this.gameSession.getGameBoard().getCellGrid());
+		      }
+			}
+			
+    private void savePuzzle(String path){
+		try
+	      {
+		
+			if(!path.toLowerCase().endsWith(".sudoku")){
+				path = path + ".sudoku";
+			}
+			
+			SavePackage savePackage = new SavePackage();
+			savePackage.setCellGrid(this.gameSession.getGameBoard().getCellGrid());
+			savePackage.setPuzzle(this.gameSession.getGameBoard().getCurrentPuzzle());
+			savePackage.setControlGrid(this.gameBoard.getControlGrid());
+	         FileOutputStream fileOut =
+	         new FileOutputStream(path);
+	         ObjectOutputStream out = new ObjectOutputStream(fileOut);
+	         
+	         out.writeObject(savePackage);
+	         out.close();
+	         fileOut.close();
+
+	      }catch(IOException i)
+	      {
+	          i.printStackTrace();
+	      }
+
+    }
+	
+	private void loadSession(SavePackage savePackage){
+		//this.gameBoard.populatePanel(gameSession.getGameBoard().getCellGrid(), gameSession);
+		
+		Puzzle puzzle = savePackage.getPuzzle();
+		
+		setMessage("");
+		DifficultyLevel difficulty;
+		if (puzzle.getDifficulty() == DifficultyLevel.Easy){
+			radEasy.setSelected(true);			
+		}
+		
+		else if (puzzle.getDifficulty() == DifficultyLevel.Medium){
+			radMedium.setSelected(true);			
+		}
+		
+		else{
+			radHard.setSelected(true);
+		}
+				
+		this.gameSession = new GameSession(puzzle);
+		this.gameSession.subscribeForCellChanges(this);
+		
+		GridPanel gameBoard = new GridPanel(savePackage.getControlGrid());
```

Comment:
This is loaded then thrown away...? I don't think it is necessary to save the JTextField[][] in the save package